LLVM upgrade

badboy opened the pull request:
As discussed in https://internals.rust-lang.org/t/need-help-with-emscripten-port/3154/46 I'm trying to update the used LLVM checkout in Rust.

I basically took @shepmaster's code and applied it on top (though I did the commits manually, the [original commits have better descriptions](https://github.com/rust-lang/rust/compare/master...avr-rust:avr-support).

With these changes I was able to build rustc. `make check` throws one last error on `run-pass/issue-28950.rs`. Output: https://gist.github.com/badboy/bcdd3bbde260860b6159aa49070a9052 

I took the metadata changes as is and they seem to work, though it now uses the module in another step. I'm not sure if this is the best and correct way.

Things to do:

* [x] ~~Make `run-pass/issue-28950.rs` pass~~ unrelated
* [x] Find out how the `PositionIndependentExecutable` setting is now used
* [x] Is the `llvm::legacy` still the right way to do these things?

cc @brson @alexcrichton

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @pnkfelix (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

badboy commented:
Ah, forgot to mention that I am using a LLVM checkout of `http://llvm.org/git/llvm.git`

frewsxcv commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -267,7 +267,7 @@ LLVMRustAddLibraryInfo(LLVMPassManagerRef PMB,
 extern "C" void
 LLVMRustRunFunctionPassManager(LLVMPassManagerRef PM, LLVMModuleRef M) {
-    FunctionPassManager *P = unwrap<FunctionPassManager>(PM);
+    llvm::legacy::FunctionPassManager *P = unwrap<llvm::legacy::FunctionPassManager>(PM);
```
`llvm::legacy::FunctionPassManager`

Tangential comment to this pull request: Does anyone know what this got replaced with since it's presumably legacy now?

eefriedman replied:
LLVM pass manager infrastructure is currently getting rewritten to be more flexible.  See https://github.com/llvm-mirror/llvm/blob/d1aa4ea020d932ffc6a3c3e4d1bbab659391c725/tools/opt/NewPMDriver.cpp#L50 for a usage example for the new API.  That said, the rewrite isn't complete yet.

alexcrichton replied:
Looks like clang is also still using this, so we probably don't need to update *just* yet until they do at least.

shepmaster commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -188,7 +188,7 @@ LLVMRustCreateTargetMachine(const char *triple,
 
     TargetOptions Options;
-    Options.PositionIndependentExecutable = PositionIndependentExecutable;
+    //Options.PositionIndependentExecutable = PositionIndependentExecutable;
```
I'm pretty sure this is a hack that only makes sense in my fork (there's no PIE for microcontrollers :-)).

Even if we wanted to remove this (doubtful), it should be removed and not commented-out. I believe that LLVM applies this flag at a more granular level now, so looking downward for another flag to apply it to seems reasonable.

badboy replied:
Right, this is one of the things mentioned above. I left it in, so I remember to fix that.

alexcrichton replied:
This was [deleted here](http://reviews.llvm.org/D19753) which appears to be [replaced by this](http://reviews.llvm.org/D19671) which is a new `setPIELevel` function on the LLVM module itself.

The LLVM module isn't available at this time so we'll just have to arrange for it to be set later. Perhaps something like this though:

* Leave this block, but gated by `#if LLVM_VERSION_MINOR <= 8` for compatibility with older versions.
* Add `LLVMRustSetModulePIELevel` as a function which is a noop on pre-LLVM-3.9 and afterwards calls the `setPIELevel` function on the LLVM module.

badboy replied:
Did you mean we have a `LLVMRustSetModulePIELevel` with the `#ifdef` for calling the `setPIELevel` method?
If so, at what point do we call it?

shepmaster replied:
> If so, at what point do we call it?

I'd believe you want to call it every time an LLVM module is created.

shepmaster commented on src/librustc_trans/debuginfo/mod.rs:
```diff
@@ -168,7 +168,7 @@ pub fn finalize(cx: &CrateContext) {
 
     debug!("finalize");
-    let _ = compile_unit_metadata(cx);
+    //let _ = compile_unit_metadata(cx);
```
Shouldn't commit commented-out code. This appears to be moved to the constructor, so I believe it's safe and correct to just remove this.

badboy replied:
Right. This will be removed.

shepmaster replied:
Although thinking of what @alexcrichton was saying, it might have to just be placed in an `#if` guard to maintain compatibility with earlier LLVM...

alexcrichton replied:
Yeah we may want this for compatibility with older llvm versions, but I'm not sure if the fix here is backwards compatible (it may be?)

badboy replied:
It will definitely need to change, as we changed the function.
However, the new version might even work with old LLVM, doesn't it?

alexcrichton replied:
I think maybe yeah? I'm not actually sure what changed here, it looks like just a bit of shuffling around to me...

shepmaster replied:
It's indeed possible that the updated code (that creates the metadata early) will work with older LLVMs. Specifically, LLVM inverted the way the debug information graph is built. All the builders inside LLVM magically handle this, **as long as there is something to add the debug information to**. Creating the debug information as early as possible enables this.

shepmaster commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -359,7 +359,17 @@ LLVMRustAddAlwaysInlinePass(LLVMPassManagerBuilderRef PMB, bool AddLifetimes) {
     llvm::legacy::PassManager passes;
-    passes.add(llvm::createInternalizePass());
+
+    auto PreserveFunctions = [=](const GlobalValue &GV) {
```
How did this work before? `symbols` wasn't even being used...

badboy replied:
It was used in the old version, see https://github.com/rust-lang/rust/pull/34743/commits/0763041c3589c042ed6c57ad49951eb7d584f675#diff-2030f049f8df2454e3720ea9403bc14aL363.

I should have squashed that.

shepmaster commented:
I'm no core contributor, but I don't see the value in commit X introducing less-than-great variable names and then commit X+1 improving them. I'd just squash those two together.

badboy commented:
@shepmaster yes, I should have squashed them beforehand. I can still do this.

alexcrichton commented on src/librustc_trans/debuginfo/metadata.rs:
```diff
@@ -981,14 +981,14 @@ fn pointer_type_metadata<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
-    let work_dir = &cx.sess().working_dir;
-    let compile_unit_name = match cx.sess().local_crate_source_file {
-        None => fallback_path(cx),
+pub fn compile_unit_metadata(scc: &::context::SharedCrateContext, debug_context: &CrateDebugContext, sess: &::session::Session) -> DIDescriptor {
```
Right now our line limit is 100 characters so I think this will trigger `make tidy`, could you also add imports instead of using absolute paths here?

alexcrichton commented on src/rustllvm/PassWrapper.cpp:
```diff
@@ -358,9 +358,18 @@ LLVMRustAddAlwaysInlinePass(LLVMPassManagerBuilderRef PMB, bool AddLifetimes) {
+        return false;
+    };
+
+    passes.add(llvm::createInternalizePass(PreserveFunctions));
```
Can you wrap this in a `#if` to only execute this new part on LLVM version >= 9 and keep the old code around for 8 and below? That should help us keep compiling against older versions of LLVM.

Come to think of that, can you also ensure that the `llvm::legacy::` above compiles on LLVM 3.7 and 3.8?

alexcrichton commented on src/llvm:
```diff
@@ -1 +1 @@
-Subproject commit 80ad955b60b3ac02d0462a4a65fcea597d0ebfb1
+Subproject commit c90294dc24393d973ed18bbc8bc87b73cb67e484
```
This is currently the [rust-llvm-2016-03-13 branch](https://github.com/rust-lang/llvm/commits/rust-llvm-2016-03-13) in the rust-lang/llvm repo, we'll want to make a new branch which points to this LLVM version and also has some of our custom patches, which includes.

I've done this at the [rust-lang-2016-07-09 branch](https://github.com/rust-lang/llvm/commits/rust-llvm-2016-07-09) as well as including some of our own patches (mostly just related to compiling LLVM itself). Can you update this commit to that patch?

alexcrichton commented:
A failure on issue-28950 looks like it may be specific to your system or something like that, I doubt that the LLVM changes would cause it to fail. Are you sure that it succeeds before this PR on your local computer?

badboy commented:
@alexcrichton Indeed, with a fresh checkout of rust-lang/rust I see the same error on issue-28950

alexbool commented:
This is supposed to fix #34119 ?

badboy commented:
@alexbool If it is fixed by something in LLVM upstream, then yes

brson commented:
cc @kripken Here's Rust's LLVM upgrade.

kripken commented:
@brson: cool, emscripten's merges of that same LLVM commit are in https://github.com/kripken/emscripten-fastcomp/tree/next-merge and https://github.com/kripken/emscripten-fastcomp-clang/tree/next-merge .

The tracking issue is https://github.com/kripken/emscripten/issues/4430. Looks like only some SIMD issues block it, which probably wouldn't be a problem for rust, so it could be usable for testing already.

shepmaster commented:
I'd strongly encourage having links to the original LLVM issues / commits that are forcing these changes. These links should be in the commit message that modifies the relevant aspect of the code. That should make it much easier for people in the future to track down the *why*.

shepmaster commented:
> That should make it much easier for people in the future to track down the why.

I say this having been one of those people to track down such information in the past.

alexcrichton commented:
@shepmaster to clarify, you mean the changes to LLVM bindings and why they're changing? (e.g. [this](https://github.com/rust-lang/rust/pull/34743#discussion_r70176713))

shepmaster commented:
> you mean the changes to LLVM bindings and why they're changing

Yeah; sorry to be unclear. Having the commit message be something like

> [LLVM 3.9] Configure PIE at the module level instead of compilation unit level
> 
> This was deleted here[linky] which appears to be replaced by this[linky] which is a new setPIELevel function on the LLVM module itself. 

would make that commit 💯 in my book. Otherwise the commit is like "moving some functions around!"

Again, I stress that I hold no power to enforce these suggestions; I just would have found commits like this to be useful when I was digging into the existing code to make these kind of changes. 🌴

alexcrichton commented:
Sounds like a good plan to me!

badboy commented:
I will rebase the commits and change accordingly (force-push incoming!)

solson commented:
@alexcrichton @eddyb Can this PR also be placed under the "Launch MIR into Orbit" milestone, since it's blocking https://github.com/rust-lang/rust/issues/34119?

alexcrichton commented:
@solson sure!

badboy commented:
I just force-pushed with the new commits

* Reorganized all commits into logical blocks, including proper descriptions and links
* I had to upgrade compiler-rt as well to get it to compile, this is not yet reflected here
* I had to re-enable the `std::thread` thing in LLVM to get it to compile at least once. We probably need to find a proper workaround first.
* When choosing anything other than `Default` in [ffc5a8608f1418f6a52f754232ba5edb90427370](https://github.com/rust-lang/rust/pull/34743/commits/ffc5a8608f1418f6a52f754232ba5edb90427370#diff-2030f049f8df2454e3720ea9403bc14aR421) I get the following error on compile:

```
error: linking with `cc` failed: exit code: 1
note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.0.o" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.1.o" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.2.o" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.3.o" "-o" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/libstd-b8d6c224bdd7e5ff.so" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.metadata.o" "-nodefaultlibs" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/build/std-26167b16044fc6c3/out/.libs" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/build/alloc_jemalloc-0986af1f8480d6ae/out/lib" "-L" "/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/libcollections-af9a9d510058b25a.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/libpanic_unwind-17c382f05cff5951.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/liballoc-3e8adf33e113f98f.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/liballoc_jemalloc-565df9f061d73123.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/librustc_unicode-f4dd78d2f1b9c500.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/libunwind-f70560ece060458e.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/liblibc-b8c0e2a7b980b655.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/librand-f0667cf9d06dbb0e.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "/tmp/rustc.FqfLoqjnVH7z/libcore-d7d9994298ae504c.rlib" "-Wl,--no-whole-archive" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "util" "-shared" "-Wl,-rpath,$ORIGIN/../lib" "-l" "compiler-rt"
note: /usr/bin/ld: /home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.0.o: relocation R_X86_64_TPOFF32 against `_ZN3std11collections4hash3map11RandomState3new4KEYS7__getit5__KEY17hc8a582c36b9a9c3eE' can not be used when making a shared object; recompile with -fPIC
/home/jer/src/rust/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/debug/deps/std-b8d6c224bdd7e5ff.0.o: error adding symbols: Bad value
collect2: error: ld returned 1 exit status

error: aborting due to previous error
error: Could not compile `std`.
```

It seems something is not compiled correctly with `-fPIC`, but I'm not sure what to do here.

brson commented:
Thanks @badboy. This is a great looking branch.

badboy commented:
Btw, I tested with commit 14f04db1b53e21922728767b440a44741c9e292c from upstream compiler-rt.

alexcrichton commented on src/librustc_trans/context.rs:
```diff
@@ -322,6 +322,7 @@ unsafe fn create_context_and_module(sess: &Session, mod_name: &str) -> (ContextR
     let llvm_target = sess.target.target.llvm_target.as_bytes();
     let llvm_target = CString::new(llvm_target).unwrap();
     llvm::LLVMRustSetNormalizedTarget(llmod, llvm_target.as_ptr());
+    llvm::LLVMRustSetModulePIELevel(llmod);
```
This is actually currently conditional when the argument is passed to `LLVMRustCreateTargetMachine`:

```rust
!any_library && reloc_model == llvm::RelocPIC
```

Perhaps that logic could be extracted to a method and that could be passed down here as well?

badboy replied:
Adding the exact same code to calculate `any_library` and `reloc_model` and then gating that call seems to work now, even with `PIELevel::Level::Large` (which, to my understanding should be equivalent to the old option).

badboy replied:
I will refactor the code into a method, so it can be called where necessary

alexcrichton replied:
Ok, awesome! So long as we're doing the same as before (whatever that PIELevel is), sounds good to me

shepmaster replied:
> whatever that PIELevel is

All I know is that *my* PIELevel is always too low. 🍰

alexcrichton commented on src/librustc_trans/debuginfo/metadata.rs:
```diff
@@ -1015,19 +1019,19 @@ pub fn compile_unit_metadata(cx: &CrateContext) -> DIDescriptor {
 
-    fn fallback_path(cx: &CrateContext) -> CString {
-        CString::new(cx.link_meta().crate_name.clone()).unwrap()
+    fn fallback_path(scc: &::context::SharedCrateContext) -> CString {
```
one more absolute import here

badboy replied:
Oh right, will fix it.

alexcrichton commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -150,19 +158,33 @@ LLVMRustWriteArchive(char *Dst,
-#if LLVM_VERSION_MINOR >= 8
-      Members.push_back(NewArchiveIterator(Member->filename));
+#if LLVM_VERSION_MINOR >= 9
+      Expected<NewArchiveMember> MOrErr = NewArchiveMember::getFile(Member->filename, true);
```
I swear this changes literally every single LLVM release

badboy replied:
I'm sure they find the best API eventually

alexcrichton commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -150,19 +158,33 @@ LLVMRustWriteArchive(char *Dst,
     assert(Member->name);
     if (Member->filename) {
-#if LLVM_VERSION_MINOR >= 8
-      Members.push_back(NewArchiveIterator(Member->filename));
```
I think this needs to stick around for LLVM 3.8, right?

badboy replied:
Hm, I'll check that again, testing it with 3.8 as well

alexcrichton replied:
FWIW in the past I've used the precompiled 3.7/3.8 distributions that LLVM ships and I just have a few build directories when I test out these changes, I just double check the C++ compiles on each and shouldn't take too long.

alexcrichton commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -150,19 +158,33 @@ LLVMRustWriteArchive(char *Dst,
                      bool WriteSymbtab,
                      Archive::Kind Kind) {
-  std::vector<NewArchiveIterator> Members;
+  std::vector<NewArchiveMember> Members;
```
This'll probably need a #ifdef for 3.7-8 compat

badboy replied:
Oops, right.

alexcrichton commented:
> When choosing anything other than Default in ffc5a8608f1418f6a52f754232ba5edb90427370 I get the following error on compile:

That seems fine to me, we probably want `Default` for now anyway. 

> I had to upgrade compiler-rt as well to get it to compile, this is not yet reflected here

Do you have a branch which contains [our patches as well](https://github.com/rust-lang/compiler-rt/commits/rust-2016-06-16)? If so, I'll push it up to rust-lang! We unfortunately have a lot of patches here :(

> I had to re-enable the std::thread thing in LLVM to get it to compile at least once.

Drat, could you send a PR to the LLVM branch in rust-lang/llvm though?

badboy commented:
> Do you have a branch which contains our patches as well? If so, I'll push it up to rust-lang! We unfortunately have a lot of patches here :(

I will try to combine those.

> Drat, could you send a PR to the LLVM branch in rust-lang/llvm though?

I can, but this will break mingw-w64 then again.

alexcrichton commented:
Ah yeah and if you have difficulty merging the compiler-rt patches, I can also help out rebasing those. Also yeah we need to get this working on mingw-w64, so were the patches to *enable* usage of `std::thread`? If so we just need to figure out how to get it working there.

Not sure how to deal with upstream adding more of what we need to remove...

badboy commented:
New commits tackle the latest comments. I tried with LLVM 3.8 and stage1 still compiles. For some reason the LLVM 3.7 fails due to changed arguments of a function called in `RustWrapper.cpp`.

Next on: compiler-rt merging

alexcrichton commented:
Well this is gonna get interested, wanted to test some things out locally and got:

```
CMake 3.4.3 or higher is required.  You are running version 3.2.2
```

I'm on Ubuntu 14.04 (which a bunch of our builders are as well), so we'll have to resolve this one way or another. We'll also need to update the README I believe to reflect that 3.4.3 is the minimum version, not 2.8.8 any more

badboy commented:
Oh right, that cmake thing. D'oh!

alexcrichton commented:
Oh dear, do we *also* need to require a super-new clang now? I'm getting:

```
//home//alex//code//clang+llvm-3.8.0-x86_64-linux-gnu-ubuntu-14.04//include/llvm/Support/Format.h:79:7: error:
      'llvm::format_object<unsigned int>' has virtual functions but non-virtual destructor
```

My clang version is `Ubuntu clang version 3.4-1ubuntu3`

alexcrichton commented:
Oh wait no nevermind that's just because `-Wnon-virtual-dtor` is in `llvm-config --cxxflags` from those pre built binaries, disregard!

alexcrichton commented:
Ok, confirmed 3.8 at least compiles, and 3.7.1 compiles for me  as well. Note that 3.7.0 doesn't compile for known reasons, so maybe you're testing against that locally? If travis passes though we should be fine.

badboy commented:
D'oh! Ok, it is 3.7.1 breaking, thanks Ubuntu…

alexcrichton commented:
Ok, CMake versions are:

```
$ docker run --entrypoint bash -it alexcrichton/rust-slave-linux-cross:2016-06-21 -c 'cmake --version'
cmake version 3.2.2
$ docker run --entrypoint bash -it alexcrichton/rust-slave-dist:2016-05-31 -c 'cmake --version'
cmake version 3.3.2
$ docker run --entrypoint bash -it alexcrichton/rust-slave-android:2016-07-01 -c 'cmake --version'
cmake version 3.5.1
$ docker run --entrypoint bash -it alexcrichton/rust-slave-linux:2016-03-30 -c 'cmake --version'
cmake version 3.2.2
$ ssh macstadium-prod1 /usr/local/bin/cmake --version
cmake version 3.5.0
```

And the Windows bots are running cmake 3.2.1

Hurray! Almost everything needs an upgrade! The linux bots should all be easy enough as they're just tweaking some docker images. Most of them are on ubuntu 14.04 for no particular reason, and Android which is on 16.04 shows that it should be far enough forward. The Windows bots will be... interesting, but I'll figure it out.

As soon as this is ready to go we can r+ and confirm it actually needs an upgrade, but I'll start upgrading them anyway

badboy commented:
Btw, this is the error I get with the current compiler-rt:

```
CMake Error at cmake/Modules/AddCompilerRT.cmake:1 (include):
  include could not find load file:

    AddLLVM
Call Stack (most recent call first):
  lib/CMakeLists.txt:4 (include)
```

It's missing some files, upstream checkouts has these files.
I'm not sure why it works in master right now.
Maybe it is enough if we place the right files there instead of doing a full compiler-rt upgrade?

alexcrichton commented:
Hm that means that build system changes probably happened, in *theory* a compiler-rt upgrade should be pretty easy, so I think it'd be worth trying that out first

badboy commented:
If you say so, I will try that first in the morning.

badboy commented:
I have a [branch of compiler-rt here](https://github.com/badboy/compiler-rt/tree/upstream-update) where I merged master from http://llvm.org/git/compiler-rt and fixed the conflicts. Right now it compiles on my test server, but I might have broken other stuff.

alexcrichton commented:
Ok cool, thanks! I'll try to go through that soon. Unfortunately it was a bit rocky updating the CMake version. So in updating the linux-cross image for Ubuntu 16.04 it means we couldn't use our old "vendored" MIPS compilers, we could now switch to the "official" MIPS compilers. Unfortunately they don't support compiling with `-msoft-float` [which we pass](https://github.com/rust-lang/rust/blob/master/mk/cfg/mips-unknown-linux-gnu.mk#L11). It looks like [not passing them](https://github.com/brson/rust/commit/7eae8b691f410e886238478771f81acb3476d4f2), however, is gonna make it [past all the bots](http://54.176.156.253/grid) (test are still running).

cc @japaric, do you have an idea of how crucial it is to pass `-msoft-float` for the MIPS target? I can't imagine we have all that many users of the `mips-unknown-linux-gnu` target right now as we haven't even had to worry about glibc compatibility just yet, but you never know!

Note that the `mipsel` target doesn't pass `-msoft-float`, so I'm tempted to just not worry about it and stop passing that flag.

@badboy I'll get back to you about the compiler-rt upgrade.

shepmaster commented:
![alexcrichton and badboy as they make progress](https://cdn.meme.am/instances/500x/69464313.jpg)

shepmaster commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -150,19 +158,40 @@ LLVMRustWriteArchive(char *Dst,
                      bool WriteSymbtab,
                      Archive::Kind Kind) {
+
+#if LLVM_VERSION_MINOR >= 9
```
Perhaps these should reference the old version, like the previous set? `LLVM_VERSION_MINOR <= 8` and flip the clauses?

shepmaster commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -150,19 +158,40 @@ LLVMRustWriteArchive(char *Dst,
       Members.push_back(NewArchiveIterator(Member->filename, Member->name));
 #endif
     } else {
+#if LLVM_VERSION_MINOR >= 9
```
Ditto the flip.

alexcrichton commented:
@badboy ok it looks like the LLVM commit I gave you didn't actually build, the `std::thread` patch was indeed bad. I [tweaked it again](https://github.com/rust-lang/llvm/commit/6e665b76d2b64256a932a61ac3eeb5ce2971c185) to hopefully be more future proof this time, and I think fewer hacks are needed now that we've upgraded our OSX compilers. Can you update the llvm submodule to https://github.com/rust-lang/llvm/commit/6e665b76d2b64256a932a61ac3eeb5ce2971c185?

alexcrichton commented:
Ok, and I have an [initial stack of patches](https://github.com/rust-lang/compiler-rt/commits/rust-2016-07-14) for compiler-rt.

I experimented dropping:

* Old MIPS patches, our compilers are much newer and hopefully won't need them
* Old ARM patches, merged upstream
* Old Android patches, hopefully our compiler is new enough to not need them
* Old SJ/LJ patches, current changes may have subsumed this? Not sure (cc @vhbit) 

Other than that couldn't drop *too* much unfortunately

Still testing locally, but if you want to update this PR we can start sending it to bors.

alexcrichton commented:
Er actually, to send to the bots we have to pull in all the new images, and to do that we have to land https://github.com/rust-lang/rust/pull/34841 first, but you get my point :)

alexcrichton commented:
Fixed some minor issues with compiler-rt, should pass the linux-cross bot now and hopefully Android as well. That's where most changes were targeted, so it *should* be good to go now I think. New commit to update to is https://github.com/rust-lang/compiler-rt/commit/3c51a621bb34f31bcb76e8e582945a7e045ce11d

japaric commented:
@alexcrichton 

> cc @japaric, do you have an idea of how crucial it is to pass -msoft-float for the MIPS target?

I looked into this mips-gcc float ABI thing. Here are my notes:

- The current rust-buildbot MIPS toolchain (15.10 ppa package) uses the soft float ABI
- with that toolchain, we don't need to pass -msoft-float because it's already the default ABI

However:

- the new rust-buildbot toolchain (16.04 official package) uses the hard float ABI(!)
- I think that's why @alexcrichton is getting an error building C dependencies with -msoft-float -- these two float ABIs are incompatible and can't be linked together
- more clearly: using the -msoft-float flag you are building new C binaries (*.o) with the soft float ABI and trying to link them with the system libraries that have a hard float ABI

Next:

- On the Rust ABI side, the mips(el)-gnu backend is generating binaries with the **soft** float ABI

Therefore:

- We should continue to use a soft float ABI toolchain to avoid mixing two different ABIs i.e. we shouldn't use the 16.04 official package, which is hard float ABI, but instead use a soft-float ABI one e.g. built with crosstool-ng

---

Now, I'm actually surprised that #34841 passes on on the upgraded docker images, which have the 16.04 mips toolchain, ... I think this is because mixing these two float ABIs only generates a linker warning and not a hard error:

```
# ubuntu 16.04
$ mips-linux-gnu-gcc-5 -msoft-float foo.c
/usr/lib/gcc-cross/mips-linux-gnu/5/../../../../mips-linux-gnu/bin/ld: Warning: a.out uses -mhard-float (set by /usr/lib/gcc-cross/mips-linux-gnu/5/../../../../mips-linux-gnu/lib/../lib/crt1.o), /tmp/ccJcvBHC.o uses -msoft-float
```

Perhaps std appears to cross compile OK (albeit with linker warnings, which rustc doesn't bubble up (?)) but using that std to cross compile rust programs generates bad binaries (because of the mixed ABIs)? -- we don't really know because we don't have tests that execute cross compiled binaries. Anyway, I would advise against mixing two different float ABIs.

alexcrichton commented:
Looks like our mips targets were turned into soft float [way back when](https://github.com/rust-lang/rust/pull/9617), so I'm not sure they necessarily *need* to be soft float. @japaric do you know if there's a reason to do one or the other? It seems odd to me that the mipsel target would be hard float when the mips target isn't...

It's true though that we definitely shouldn't mix, but I think I'd personally prefer to just move everything over to hard float to match upstream toolchains if possible.

alexcrichton commented:
Oh dear, so... this will probably break external builds. Look like compiler-rt can't be built against older LLVM sources!

@brson thoughts on doing https://github.com/rust-lang/rust/issues/34400 *now*?

alexcrichton commented:
Ok, thanks again for all the patience @badboy! I think we have a plan of action to solve the remaining problems:

1. To update cmake (to build LLVM 3.9) we will update the bots to Ubuntu 16.04, which pulls in a new MIPS toolchain (all other toolchains remain the same), which as @japaric [points out](https://github.com/rust-lang/rust/pull/34743#issuecomment-233101493) has some differences in float support. I'll send a PR to change us over to hard-float for the MIPS targets (more consistent with debian), and it shouldn't affect this PR I believe.
2. To fix compiling compiler-rt against older LLVMs (or any other LLVM) we're going to pursue https://github.com/rust-lang/rust/pull/34873. This will cut our ties with compiler-rt's build system, so all these questions go away!

So the remaining steps here are:

* I will send a PR to use hard-float on MIPS, that can land whenever
* We'll land https://github.com/rust-lang/rust/pull/34873
* I'll tweak the compiler-rt submodule to jettison all our build system patches and only keep the source patches.
* You'll rebase this PR, including updates to the compiler-rt submodule
* We'll wait for a green light on Travis
* I'll restart buildbot, updating all the images
* We'll send to @bors 

Sound ok to you? Any last things you want to throw in on this branch?

badboy commented:
:+1: That sounds like a good plan (don't worry about my patience, I'm happy to get so much support in bringing this forward). I'll take a look at #34873 to understand what it does.

brson commented:
> @brson thoughts on doing #34400 now?

Yes.

alexcrichton commented:
Changing MIPS to hard float is #34910

alexcrichton commented:
Ok https://github.com/rust-lang/rust/pull/34873 is almost ready to land (fingers crossed for this round). Unfortunately this branch doesn't build for me but with https://github.com/brson/rust/commit/7f5f54ddc5eec9529aef0d266c515806671ea31c it seems to work. I had to update the `llvm-mc` syntax b/c it changed in 3.9 and there was just one modification to a C++ file we had to keep from before. Other than that things are looking good so far.

If everything goes *super* well I'll just make a new PR with all these commits, that one, and the updated compiler-rt. Then I'll r+ that and see it merged when I wake up. I can dream can't I?

alexcrichton commented:
Alas perhaps not tonight, the g++ on the MinGW buildbots refuses to build LLVM no matter how I'm coaxing it right now, so gonna just have to sort that out tomorrow.

alexcrichton commented:
@badboy can you cherry pick 31daccb33cf7e36d5e06131a5034b85a1ccc66dd and f7c8e80c8b6bd384838d8ef71423f61a311bc4f3 in here? No need to preserve commit messages or authorship or anything, they'll just need to be in this PR. The LLVM submodule will likely change again as I figure out how to compile it on MinGW again.

badboy commented:
Cherry-picks incoming :cherries:

badboy commented:
There we go, I picked them apart a bit, but the changes are the same.

alexcrichton commented:
Ok, I think I [got LLVM working](https://github.com/rust-lang/llvm/commit/68447bd81e1c7e588466b91285f1bd9d641de465) through some serious surgery. Looks like I [messed up a bit](https://github.com/brson/rust/commit/de7cc20868045357b6f2d5a5ca89674deb8e1fa1#diff-e2d5a00791bce9a01f99bc6fd613a39dL1024) with the grep condition in configure.

Also looks like tests are failing on the master branch because LLVM changed one of their custom target specifications, which we assert is the same as our own. If we change it, however, then we break with other LLVMs because they have a different data-layout as well. I'm just gonna disable that check for custom LLVMs for now and open an issue. Still working through failing tests though.

brson commented:
Wow that is a horrible patch :(

badboy commented:
@alexcrichton Should I already pull in the updates to the LLVM branch and the WIP changes?

badboy commented:
Ok, so I'm unable to find the actual commits in any cloned version of these repositories…

alexcrichton commented:
@badboy not quite yet, I'm still working on testing. Unfortunately there are tons of failures on the bots:

* [auto-linux-32-nopt-t](http://54.176.156.253/builders/auto-linux-32-nopt-t/builds/85/steps/test/logs/stdio)
* [auto-linux-64-cross-freebsd](http://54.176.156.253/builders/auto-linux-64-cross-freebsd/builds/52/steps/compile/logs/stdio)
* [auto-linux-64-debug-opt](http://54.176.156.253/builders/auto-linux-64-debug-opt/builds/63/steps/compile/logs/stdio)
* [auto-linux-cross-opt](http://54.176.156.253/builders/auto-linux-cross-opt/builds/93/steps/compile/logs/stdio)
* [auto-mac-32-opt](http://54.176.156.253/builders/auto-mac-32-opt/builds/50/steps/compile/logs/stdio)
* [auto-win-msvc-64-opt](http://54.176.156.253/builders/auto-win-msvc-64-opt-mir/builds/47/steps/compile/logs/stdio)

And... they're basically all failing for different reasons.

I had to apply [this patch](https://github.com/brson/rust/commit/40b60755ca8f5496c2e26140c4ed3f05edad9318) to get to that state, which:

* Should pass tests on linux64 (locally for me)
  * The data-layout for aarch64 was changed, the data-layout assertion was relaxed for custom LLVMs
  * A test was removed which I don't think ever actually needed to work.
* Uses an LLVm that actually compiles on MinGW
* Fixes a condition in the `configure` script.

alexcrichton commented:
I'm currently playing whack-a-mole with the bots in dev, trying to get a full suite of the failures so we can work through them.

alexcrichton commented:
Ok, I've got a full build now, failures are:

* [auto-linux-32-nopt-t](http://54.176.156.253/builders/auto-linux-32-nopt-t/builds/86)
* [auto-linux-32-opt](http://54.176.156.253/builders/auto-linux-32-opt/builds/87)
* [auto-linux-64-cross-freebsd](http://54.176.156.253/builders/auto-linux-64-cross-freebsd/builds/53)
* [auto-linux-64-debug-opt](http://54.176.156.253/builders/auto-linux-64-debug-opt/builds/64)
* [auto-linux-64-x-android-t](http://54.176.156.253/builders/auto-linux-64-x-android-t/builds/95)
* [auto-linux-cross-opt](http://54.176.156.253/builders/auto-linux-cross-opt/builds/94)
* [auto-mac-32-opt](http://54.176.156.253/builders/auto-mac-32-opt/builds/51)
* [auto-win-gnu-32-opt](http://54.176.156.253/builders/auto-win-gnu-32-opt/builds/75)
* [auto-win-gnu-32-opt-rustbuild](http://54.176.156.253/builders/auto-win-gnu-32-opt-rustbuild/builds/58)
* [auto-win-msvc-64-cargotest](http://54.176.156.253/builders/auto-win-msvc-64-cargotest/builds/52)
* [auto-win-msvc-64-opt](http://54.176.156.253/builders/auto-win-msvc-64-opt/builds/118)
* [auto-win-msvc-64-opt-mir](http://54.176.156.253/builders/auto-win-msvc-64-opt-mir/builds/48)
* [auto-win-msvc-64-opt-rustbuild](http://54.176.156.253/builders/auto-win-msvc-64-opt-rustbuild/builds/56)

alexcrichton commented:
I unfortunately won't get a chance to look into these test failures yet, so any help investigating them would be much appreciated!

eddyb commented on src/rustllvm/ArchiveWrapper.cpp:
```diff
@@ -71,7 +79,12 @@ extern "C" RustArchiveIterator*
     rai->cur = ar->child_begin();
+#else
+    Error err;
+    rai->cur = ar->child_begin(err);
```
The `err` `Error` value needs to be checked, otherwise this happens with debug LLVM:
```
Program aborted due to an unhandled Error:
Error value was Success. (Note: Success values must still be checked prior to being destroyed).
```

badboy replied:
will take care of that

eddyb commented on src/librustc_trans/context.rs:
```diff
@@ -322,6 +355,11 @@ unsafe fn create_context_and_module(sess: &Session, mod_name: &str) -> (ContextR
     llvm::LLVMRustSetNormalizedTarget(llmod, llvm_target.as_ptr());
+
+    if is_pie_binary(sess) {
+        llvm::LLVMRustSetModulePIELevel(llmod);
```
You might need to set the `PICLevel` too, even if the relocation model is `PIC`.
One potentially related failure is:
```
note: ld: illegal text-relocation to '_panic_loc21555' in i686-apple-darwin/stage1/lib/rustlib/i686-apple-darwin/lib/std-f53fb285.0.o from '__ZN103_$LT$core..slice..RSplitN$LT$$u27$a$C$$u20$T$C$$u20$P$GT$$u20$as$u20$core..iter..iterator..Iterator$GT$4next17he6277dfc64030327E' in i686-apple-darwin/stage1/lib/rustlib/i686-apple-darwin/lib/std-f53fb285.0.o for architecture i386
```

eddyb commented:
All the relocation nonsense was because, well, this mismatch:
```rust
pub enum RelocMode {
    RelocDefault = 0,
    RelocStatic = 1,
    RelocPIC = 2,
    RelocDynamicNoPic = 3,
}
```
```cpp
typedef enum {
    LLVMRelocDefault,
    LLVMRelocStatic,
    LLVMRelocPIC,
    LLVMRelocDynamicNoPic
} LLVMRelocMode;
...
enum Model { Static, PIC_, DynamicNoPIC };
```

So the Rust side had `LLVMRelocMode`, but `LLVMRustCreateTargetMachine` was taking `Reloc::Model` which meant that `RelocMode::RelocPIC` became `Reloc::DynamicNoPIC`.

The `LLVMRust*` functions **MUST ONLY TAKE C TYPES**.
In this case, `LLVMRelocMode`, and do [the conversion `LLVMCreateTargetMachine` does](https://github.com/llvm-mirror/llvm/blob/c9b262bfbd5b9fb6f10749dba1465569f39bd625/lib/Target/TargetMachineC.cpp#L104-L121).

eddyb commented:
Aarch64 Android also needs a `data-layout` update:
```
data-layout for builtin `aarch64-linux-android` target,
`e-m:e-i64:64-i128:128-n32:64-S128`, differs from LLVM default,
`e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128` 
```

badboy commented:
Thanks @eddyb, I'll take a look later today.

badboy commented:
Again, thanks @eddyb. I tracked it down to this commit: https://github.com/rust-lang/rust/commit/73540551#diff-5bb596c9f589759bef3b0b31f5505d2eR2107
It seems this was a type mismatch for quite some time.
I will fix it in the above code.

DemiMarie commented:
Would it be possible to check with an LLVM clone over HTTPS? Just to make sure that nobody has MITM'd the insecure HTTP clone.

badboy commented:
@DemiMarie not sure what you mean. The submodule in this repository is at https://github.com/rust-lang/llvm, which is fetched over https.

badboy commented:
I changed the type on `LLVMRustCreateTargetMachine` and added the conversion.
I also added explicit checks for the error, which has to be done in each iteration (did I mention that this API seems super weird o_O).
cc @eddyb 

I'm no C++ expert and I don't understand this iterator at all, so I might be doing stuff wrong

badboy commented:
Ah damn, the optional is only necessary on newer LLVM -_-

badboy commented:
I added a conditional to use the raw type. Though right now it does not check `RelocDefault`. What should be the default in LLVM <= 3.8?

eddyb commented:
@badboy The old implementation [uses `Reloc::Default`](https://github.com/rust-lang/llvm/blob/7ca76af03bb04659562890d6b4f223fffe0d748f/lib/Target/TargetMachineC.cpp#L117-L131) - btw, you can see that `LLVMCodeModel` and `LLVMCodeGenOptLevel` also need to be converted, if they match now it may be pure coincidence.

alexcrichton commented:
Ok, I've pushed this for testing.

@badboy could you also update with the fixes in https://github.com/brson/rust/commit/7af5c985826da4a4b1fbcd6fad32ed760579d861?

alexcrichton commented:
Ah it also looks like [LLVM 3.9 has been branched](http://lists.llvm.org/pipermail/llvm-dev/2016-July/102468.html), so if this passes all tests I may do one final rebase to get us onto the 3.9 release branch so we can just cherry-pick back their fixes easily if necessary.

badboy commented:
Will add the fixes.
On IRC someone mentioned me about some of the C++ changes. I'll check that tomorrow as well.

alexcrichton commented:
Ah and FWIW I've been playing around with some more commits, and what I've done so far is:

* Rebase this branch on master
* Then apply the commits at https://github.com/brson/rust/commits/f8728ac0e2b4c36e8ce9e0c9a5169b32ad5cb40e

I've updated LLVM to the 3.9 release in that branch as well, and I think it's got most tests fixed now. The remaining thing to deal with is MSVC and `llvm-config`.

alexcrichton commented:
Ok I fixed up the llvm-config parsing for real this time (at https://github.com/rust-lang/rust/commits/c0e95ccd7c2bc5cd8e91aee25d65ae2699abb735).

Unfortunately all of the MSVC builders are still failing. They're all failing to compile getopts in stage1:

[logs](http://54.176.156.253/builders/auto-win-msvc-32-opt/builds/86/steps/compile/logs/stdio)

```
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libcollections
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libpanic_abort
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libunwind
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libpanic_unwind
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libstd
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libarena
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libflate
rustc: i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/libgetopts
/c/bot/slave/auto-win-msvc-32-opt/build/mk/target.mk:212: recipe for target 'i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/stamp.getopts' failed
>> rustjob: found 0 remaining processes
make: *** [i686-pc-windows-msvc/stage1/lib/rustlib/i686-pc-windows-msvc/lib/stamp.getopts] Segmentation fault
program finished with exit code 2
elapsedTime=1672.982000
```

eddyb commented:
@alexcrichton My guess is a LLVM assertion that's not checked and leads to a segfault. It would be useful, at least here, to enable LLVM assertions.

alexcrichton commented:
Ok I sent https://github.com/rust-lang/rust/commit/c0e95ccd7c2bc5cd8e91aee25d65ae2699abb735 to bors and everything is green except MSVC, which is segfaulting.

eddyb commented:
I was wrong. It's an implementation bug, https://github.com/llvm-mirror/llvm/commit/11bf1ab28279cb5f189ac4381e3aa23efd703ea3 seems to be the fix. Rebasing on top of master should include it.

alexcrichton commented:
Ok, I've rebased LLVM and have sent that to dev-bors.

Results will be at http://54.176.156.253/grid as they come in

eddyb commented:
[MSVC seems to be crashing in a few tests](http://54.176.156.253/builders/auto-win-msvc-64-opt-rustbuild/builds/67/steps/test/logs/stdio), which are actually assertion failures in debuginfo:

```cpp
rustc: /home/eddy/Projects/rust/src/llvm/lib/CodeGen/LexicalScopes.cpp:159:
llvm::LexicalScope* llvm::LexicalScopes::getOrCreateRegularScope(const llvm::DILocalScope*):
Assertion `cast<DISubprogram>(Scope)->describes(MF->getFunction())' failed.
```
cc @michaelwoerister

badboy commented:
vlad on IRC tried a LLVM 3.9 on Windows last week or so and has some tiny patches in [his branch](https://github.com/vvuk/rust/commits/codeview). 
More or less just tiny build system fixes, except for https://github.com/vvuk/rust/commit/c18c244880495c9c04365b68fe602e9e690c8e69

Might this be a fix for above crashes?

alexcrichton commented:
Nice idea! I've sent that to bors-dev and we'll see what happens

alexcrichton commented:
Ok, looks like there may be two errors:

* [auto-linux-64-debug-opt](http://54.176.156.253/builders/auto-linux-64-debug-opt/builds/77) - I think rustdoc is spuriously segfaulting?
* [auto-win-msvc-64-cargotest](http://54.176.156.253/builders/auto-win-msvc-64-cargotest/builds/65) - hyper is segfaulting when compiling

for the latter, I got a backtrace:

```
rustc: ../../../../src/llvm/lib/CodeGen/LexicalScopes.cpp:159: llvm::LexicalScope* llvm::LexicalScopes::getOrCreateRegularScope(const llvm::DILocalScope*): Assertion `cast<DISubprogram>(Scope)->describes(MF->getFunction())' failed.

Program received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffed7db700 (LWP 17806)]
0x00007ffff7133c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56      ../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7133c37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff7137028 in __GI_abort () at abort.c:89
#2  0x00007ffff712cbf6 in __assert_fail_base (fmt=0x7ffff727d3b8 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x7ffff28170f0 "cast<DISubprogram>(Scope)->describes(MF->getFunction())", file=file@entry=0x7ffff2816fb0 "../../../../src/llvm/lib/CodeGen/LexicalScopes.cpp", 
    line=line@entry=159, 
    function=function@entry=0x7ffff28178a0 <llvm::LexicalScopes::getOrCreateRegularScope(llvm::DILocalScope const*)::__PRETTY_FUNCTION__> "llvm::LexicalScope* llvm::LexicalScopes::getOrCreateRegularScope(const llvm::DILocalScope*)") at assert.c:92
#3  0x00007ffff712cca2 in __GI___assert_fail (assertion=0x7ffff28170f0 "cast<DISubprogram>(Scope)->describes(MF->getFunction())", 
    file=0x7ffff2816fb0 "../../../../src/llvm/lib/CodeGen/LexicalScopes.cpp", line=159, 
    function=0x7ffff28178a0 <llvm::LexicalScopes::getOrCreateRegularScope(llvm::DILocalScope const*)::__PRETTY_FUNCTION__> "llvm::LexicalScope* llvm::LexicalScopes::getOrCreateRegularScope(const llvm::DILocalScope*)") at assert.c:101
#4  0x00007ffff16e3430 in llvm::LexicalScopes::getOrCreateRegularScope(llvm::DILocalScope const*) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#5  0x00007ffff16e3340 in llvm::LexicalScopes::getOrCreateRegularScope(llvm::DILocalScope const*) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#6  0x00007ffff16e3eaa in llvm::LexicalScopes::extractLexicalScopes(llvm::SmallVectorImpl<std::pair<llvm::MachineInstr const*, llvm::MachineInstr const*> >&, llvm::DenseMap<llvm::MachineInstr const*, llvm::LexicalScope*, llvm::DenseMapInfo<llvm::MachineInstr const*>, llvm::detail::DenseMapPair<llvm::MachineInstr const*, llvm::LexicalScope*> >&) ()
   from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#7  0x00007ffff16e41f9 in llvm::LexicalScopes::initialize(llvm::MachineFunction const&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#8  0x00007ffff15ed8ef in llvm::DebugHandlerBase::beginFunction(llvm::MachineFunction const*) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#9  0x00007ffff163de40 in llvm::CodeViewDebug::beginFunction(llvm::MachineFunction const*) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#10 0x00007ffff15e3ae6 in llvm::AsmPrinter::EmitFunctionHeader() () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#11 0x00007ffff15dc5f5 in llvm::AsmPrinter::EmitFunctionBody() () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#12 0x00007ffff1298250 in llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#13 0x00007ffff176d013 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#14 0x00007ffff1fa7b03 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#15 0x00007ffff1fa7f2b in llvm::FPPassManager::runOnModule(llvm::Module&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#16 0x00007ffff1fa823f in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /home/alex/code/rust2/x86_64-unknown-linux-gnu/stage1/bin/../lib/librustc_llvm-f53fb285.so
#17 0x00007ffff0c36a8d in LLVMRustWriteOutputFile (Target=0x7fffdc85de20, PMR=0x7fffdf30e920, M=0x7fffebaabbc0, path=<optimized out>, FileType=llvm::TargetMachine::CGFT_ObjectFile)
```

eddyb commented:
@alexcrichton It's segfaulting in `rustc` while *compiling* `librustc`, not sure how that could be spurious.
Could be a LLVM assertion though. If *only* we had those enabled always, we could tell :).

The MSVC one seems identical to the more general debuginfo problem which was supposedly fixed.

eddyb commented on src/librustc_trans/debuginfo/metadata.rs:
```diff
@@ -68,7 +68,6 @@ pub const UNKNOWN_LINE_NUMBER: c_uint = 0;
 
 // ptr::null() doesn't work :(
-pub const NO_FILE_METADATA: DIFile = (0 as DIFile);
 pub const NO_SCOPE_METADATA: DIScope = (0 as DIScope);
```
Is it possible `NO_SCOPE_METADATA` is also wrong now?

badboy replied:
You might be write. As far as I traced it the passed value gets `unwrap`ped (which simply does a `reinterpret_cast`). Unwrapping a non-existent value in C++ is probably the same kind of bad as it is in Rust, except that in Rust it's properly safe to do so and `panic`.

michaelwoerister commented:
> MSVC seems to be crashing in a few tests, which are actually assertion failures in debuginfo...

I've seen this kind of assertion in the past in the following cases:
1. When function calls where not assigned a debug location. Then inlining would lead to instructions seemingly coming from another function.
2. When a bug lead to closures being translated multiple times, leading to multiple debuginfo descriptions for the same LLVM function.

But this seems more likely to be a change in what kind of values LLVM expects in its API?

alexcrichton commented:
As an update on this I ran against @bors last night with [this tree](https://github.com/brson/rust/commits/ebd64f2154f42ed0d60857337d66ad548f053810) which [included an update](https://github.com/rust-lang/llvm/commit/327e422d9b9df5bc02d947b69d8020817b5a2bfc) to the LLVM submodule that @eddyb was able to track down. This fixes the regression of hyper on auto-win-msvc-64-cargotest.

The final remaining failure is:

* [auto-linux-64-debug-opt](http://54.176.156.253/builders/auto-linux-64-debug-opt/builds/78)

```
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_passes
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_save_analysis
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_typeck
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_mir
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_trans
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_plugin
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_borrowck
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_driver
rustc: x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustdoc
Segmentation fault (core dumped)
make: *** [x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustdoc] Error 139
/buildslave/rust-buildbot/slave/auto-linux-64-debug-opt/build/mk/target.mk:212: recipe for target 'x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/stamp.rustdoc' failed
program finished with exit code 2
elapsedTime=10602.248647
```

@badboy Can you include a few more commits here as well?

* https://github.com/brson/rust/commit/29c9cb3d90a1c236525695cca85ab0f1d450429f - a fix to an MSVC test
* https://github.com/brson/rust/commit/be0416b89061aa7ab6244b426cca0e4de3151156 - fix for the ArchiveWrapper compilation
* https://github.com/brson/rust/commit/ebd64f2154f42ed0d60857337d66ad548f053810 - the most recent LLVM update

bors commented:
:umbrella: The latest upstream changes (presumably #34956) made this pull request unmergeable. Please resolve the merge conflicts.

badboy commented:
I just rebased this PR on master, applied the remaining patches and pushed it again.

eddyb commented:
@bors r+ p=10

bors commented:
:pushpin: Commit 7c0cd30 has been approved by `eddyb`

<!-- @bors r=eddyb 7c0cd30c4b49905cb9ddf83daa3acfc77906ff91 -->

bors commented:
:hourglass: Testing commit 7c0cd30 with merge 554bc72...

bors commented:
:broken_heart: Test failed - [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5137)

alexcrichton commented:
@bors: retry

* forgot to update the windows bots, now doing so

alexcrichton commented:
@bors: force

alexcrichton commented:
@bors: retry

alexcrichton commented:
@bors: force

michaelwoerister commented:
Use the Force, young Skywalker!

alexcrichton commented:
@bors: retry force clean

bors commented:
:hourglass: Testing commit 7c0cd30 with merge bea0adf...

bors commented:
:broken_heart: Test failed - [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10020)

alexcrichton commented:
@bors: retry force clean

bors commented:
:hourglass: Testing commit 7c0cd30 with merge 21ea9cd...

bors commented:
:broken_heart: Test failed - [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1966)

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit 7c0cd30 with merge 8f92cbd...

bors commented:
:broken_heart: Test failed - [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3307)

badboy commented:
Hmpf. Is there anyway to get anymore debugable info out of that build?

eddyb commented:
@badboy We've tried. Preloading `libSegFault.so` doesn't work for some reason and Ubuntu has that stupid apport thing that interferes with core dumps. @alexcrichton would have to fix the apport nonsense on the bots before triggering a full build with all the bots on, it's harder to reproduce in isolation.

tmiasko commented:
I had similar segfault while building this branch. rustc consistently crashed
while deleting DIBuilder from LLVM. Starting from a clean working directory
fixed it. Given that buildbot reports that "clean-llvm cleaned llvm skipped",
maybe it could also help here.

eddyb commented:
@tmiasko That's interesting. Do you have a backtrace? We couldn't get anything, reproduction seemed to need the concurrent builds, or a specific builder, to trigger.

alexcrichton commented:
@tmiasko did some digging and discovered that https://reviews.llvm.org/D22858 may be relevant ([stack trace gathered](https://pastebin.mozilla.org/8888518)). I've updated our LLVM submodule with this commit (rebased on the llvm/release_39 branch) and now it's rust-lang/llvm@d1cc489. @badboy wanna try updating to that and give it a go? Builds in dev are running, but we can do in parallel hopefully :)

badboy commented:
Done! Upgraded to d1cc489

eddyb commented:
@bors r+

bors commented:
:pushpin: Commit 5d1d247 has been approved by `eddyb`

<!-- @bors r=eddyb 5d1d2475232d06b2a315d87481898819bb547f97 -->

bors commented:
:hourglass: Testing commit 5d1d247 with merge 2c1612c...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/10066), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/10113), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/1356), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/1182), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/3333), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/10051), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/10022), [auto-linux-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/1617), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1995), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9999), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/3223), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/5173), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/10089), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/10069), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/2016), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/1192), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/5173), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/2044), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/5158), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/4347), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/1374), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/5158), [auto-win-msvc-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/1652), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/2038)
Approved by: eddyb
Pushing 2c1612c62aa59e40cf1a4bddde943938e0147eec to master...

badboy commented:
Wuhu! 🎉

eminence commented:
Outstanding work everyone!

michaelwoerister commented:
Epic! `:)`

shepmaster commented:
Hurrah! Now it's time to rebase the AVR and Emscripten forks and really give it a workout ;-)

badboy commented:
@shepmaster: That's the next step and I'll try to work on it this week

bstrie commented:
Heroes, every one of you! On with MIR!

whitequark commented:
Looks like this missed the LLVM version check in configure, not sure why it didn't break bots:
```
--- .../rust/configure	Sat Aug 13 01:03:47 2016
+++ .../rust/configure	Sat Aug 13 01:03:50 2016
@@ -1013,7 +1013,7 @@
     LLVM_VERSION=$($LLVM_CONFIG --version)
 
     case $LLVM_VERSION in
-        (3.[7-8]*)
+        (3.[7-9]*)
             msg "found ok version of LLVM: $LLVM_VERSION"
             ;;
         (*)
```

shepmaster commented:
I have a PR out for that - #35594. No idea about the bots.
